parlai/core/torch_image_agent.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""
+Subclass of TorchAgent used for handling image features.
+"""
+
+from abc import abstractmethod
+from typing import List
+
+import torch
+
+from parlai.core.message import Message
+from parlai.core.torch_agent import Batch, TorchAgent
+
+
+class TorchImageAgent(TorchAgent):
+    """
+    Subclass of TorchAgent that allows for encoding image features.
+
+    Provides flags and utility methods.
+    """
+
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        """
+        Add command-line arguments specifically for this agent.
+        """
+        super(TorchImageAgent, cls).add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('Image args')
+        agent.add_argument(
+            '--image-features-dim',
+            type=int,
+            default=2048,
+            help='Dimensionality of image features',
+        )
+        agent.add_argument(
+            '--image-encoder-num-layers',
+            type=int,
+            default=1,
+            recommended=1,
+            help='Number of linear layers to encode image features with',
+        )
+        return agent
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        self.image_features_dim = opt['image_features_dim']
+        self.image_encoder_num_layers = opt['image_encoder_num_layers']
+
+    def batchify(self, obs_batch: List[Message], sort: bool = False) -> Batch:
+        """
+        Override to handle image features.
+        """
+        batch = super().batchify(obs_batch, sort)
+        batch = self.batchify_image_features(batch)
+        return batch
+
+    @abstractmethod
+    def batchify_image_features(self, batch: Batch) -> Batch:
+        """
+        Put this batch of images into the correct format for this agent.
+
+        self._process_image_features() will likely be useful for this.
+        """
+        raise NotImplementedError(
+            'Subclasses must implement method for batching images!'
```

Comment:
is there not any shared functionality that could go here?

Reply:
Perhaps - that's what I've created https://github.com/facebookresearch/ParlAI/issues/2461 to solve. I can try to look into unifying the `batchify_image_features()` methods now if it's really important, but given that there may be image_seq2seq edge cases that we have to account for, my preference would be to make this a separate PR